api_tests/nodes/test_views.py

```diff
@@ -3411,3 +3431,233 @@ def test_node_link_already_exists(self):
+        assert_equal(res.json['data']['attributes']['content'], self.payload['data']['attributes']['content'])
+
+    def test_private_node_logged_in_read_contributor_can_comment(self):
+        res = self.app.post_json_api(self.private_url, self.payload, auth=self.read_only_contributor.auth)
+        assert_equal(res.status_code, 201)
+        assert_equal(res.json['data']['attributes']['content'], self.payload['data']['attributes']['content'])
+
+    def test_private_node_non_contributor_cannot_comment(self):
+        res = self.app.post_json_api(self.private_url, self.payload, auth=self.non_contributor.auth, expect_errors=True)
+        assert_equal(res.status_code, 403)
+
+    def test_private_node_logged_out_user_cannot_comment(self):
+        res = self.app.post_json_api(self.private_url, self.payload, expect_errors=True)
+        assert_equal(res.status_code, 401)
+
+    def test_private_node_with_public_comment_level_non_contributor_cannot_comment(self):
+        """ Test non-contributors cannot comment on a private project
+            with comment_level == 'public' """
+        project = ProjectFactory(is_public=False, creator=self.user)
+        project.comment_level = 'public'
+        project.save()
+        url = '/{}nodes/{}/comments/'.format(API_BASE, project._id)
+        res = self.app.post_json_api(url, self.payload, auth=self.non_contributor.auth, expect_errors=True)
+        assert_equal(res.status_code, 403)
+
+    def test_public_node_any_logged_in_user_can_comment(self):
+        """ Test non-contributors can comment on a public project when it is
+            configured so any logged-in user can comment (comment_level == 'public') """
+        res = self.app.post_json_api(self.public_comments_url, self.payload, auth=self.non_contributor.auth)
+        assert_equal(res.status_code, 201)
+        assert_equal(res.json['data']['attributes']['content'], self.payload['data']['attributes']['content'])
+
+        res = self.app.post_json_api(self.public_comments_url, self.payload, expect_errors=True)
+        assert_equal(res.status_code, 401)
+
+    def test_public_node_only_contributors_can_comment(self):
+        """ Test only contributors can comment on a public project when it is
+            configured so only contributors can comment (comment_level == 'private') """
+        res = self.app.post_json_api(self.public_url, self.payload, auth=self.user.auth)
+        assert_equal(res.status_code, 201)
+        assert_equal(res.json['data']['attributes']['content'], self.payload['data']['attributes']['content'])
+
+        res = self.app.post_json_api(self.public_url, self.payload, auth=self.read_only_contributor.auth)
+        assert_equal(res.status_code, 201)
+        assert_equal(res.json['data']['attributes']['content'], self.payload['data']['attributes']['content'])
+
+        res = self.app.post_json_api(self.public_url, self.payload, auth=self.non_contributor.auth, expect_errors=True)
```

Comment:
Where possible, try to keep separate scenarios in separate test methods. In addition to making sure that one failure doesn't mask others, this makes sure that tests run independently, and don't fail for unexpected reasons. (eg if the first success changed the state of the project in some way that affected things downstream)

Please separate these three cases out into separate tests. As it stands, if one assertion fails, several other independent situations may never be run.